lib/wincred/wincred.rb

```diff
@@ -0,0 +1,226 @@
+
+      enum_result = Native.CredEnumerateW(
+        nil,
+        Native::FLAGS_NONE,
+        cred_count_ptr,
+        pp_credentials.ref
+      )
+
+      if enum_result.zero?
+        raise "Enumerate credentials in WinCred failed. Error code: #{Native.GetLastError}" 
+      end
+
+      # Convert count bytes to Ruby int
+      count = cred_count_ptr[0, 4].unpack('L').first
+
+      # Convert memory struct array into Ruby array
+      count.times.map do |ndx|
+        cred_ptr = (pp_credentials + (ndx * Fiddle::SIZEOF_VOIDP)).ptr
+        cred = Native::CREDENTIALW.new(cred_ptr)
+        cred_to_hash(cred)
+      end
+    ensure
+      # Free the credential memory
+      Native.CredFree(pp_credentials) if pp_credentials
+    end
+
+    def write_credential(
+        target:,
+        username:,
+        value:
+      )
+      # Convert the unicode string to a byte array
+      value_blob = value.encode('utf-8').unpack('c*')
+
+      cred = Native::CREDENTIALW.malloc
+
+      cred.Flags = Native::FLAGS_NONE
+      cred.Type = Native::CRED_TYPE_GENERIC
+      cred.TargetName = target.encode('utf-16le')
+      cred.Comment = ''
+      cred.CredentialBlobSize = value_blob.size
+      cred.CredentialBlob = Fiddle::Pointer[value_blob.pack('c*')]
+      cred.Persist = Native::CRED_PERSIST_LOCAL_MACHINE
+      cred.AttributeCount = 0
+      cred.Attributes = nil
+      cred.TargetAlias = nil
+      cred.UserName = username.encode('utf-16le')
+
+      write_result = Native.CredWriteW(Fiddle::Pointer[cred], 0)
+
+      raise "Write to WinCred failed. Error code: #{Native.GetLastError}" if write_result.zero?
+    ensure
+      Native.CredFree(cred) if cred
+    end
+
+    def read_credential(target)
+      cred_ptr = Fiddle::Pointer.new 0
+
+      # We have to encode the target name for the Windows API call
+      read_result = Native.CredReadW(
+        target.encode('utf-16le'),
+        Native::CRED_TYPE_GENERIC,
+        Native::FLAGS_NONE,
+        cred_ptr.ref
+      )
+
+      raise "Read from WinCred failed. Error code: #{Native.GetLastError}" if read_result.zero?
+
+      cred = Native::CREDENTIALW.new(cred_ptr)
+      cred_to_hash(cred)
+    ensure
+      Native.CredFree(cred_ptr) if cred_ptr
+    end
+
+    def delete_credential(target)
+      delete_result = Native.CredDeleteW(
+        target.encode('utf-16le'),
+        Native::CRED_TYPE_GENERIC,
+        Native::FLAGS_NONE
+      )
+
+      raise "Delete from WinCred failed. Error code: #{Native.GetLastError}" if delete_result.zero?
+    end
+
+    private
+
+    def cred_to_hash(cred)
+      {
+        target: read_wide_string(cred.TargetName),
+        username: read_wide_string(cred.UserName),
+        value: (cred.CredentialBlobSize > 0 ? cred.CredentialBlob.to_s : nil)
+      }
+    end
+
+    def read_wide_string(ptr)
+      idx = 0
+      str = ''
+      while ptr[idx, 2] != "\0\0"
+        str += ptr[idx, 2].force_encoding('UTF-16LE').encode('UTF-8')
+        idx += 2
+      end
```

Comment:
The problem I'm trying to overcome is that `ptr.to_str` doesn't work at all for the double wide strings. It only returns the first character of the string contained in the pointer.

I'm not knowledgable enough in string encoding to really know how to address your concern. I will do some additional learning, but do you also have any additional suggestions in mind?

Reply:
Well, I admit it's mostly of an theoretical importance really, since it's unlikely a legit credential would have characters out of the Basic Multilingual Plane, but still it'd be good to acknowledge the deficiency in a comment at least :)

But anyway, here's what I came up with:
```ruby
def wcsnlen ptr, maxlen = 1024
  (0...maxlen).find{|i| (ptr[i*2] | ptr[i*2 + 1]) == 0 }
end

def pwchar2str ptr
  ptr.to_str(wcsnlen(ptr) * 2).force_encoding('utf-16le')
end